apps/designer/app/shared/tree-utils.ts

```diff
@@ -100,6 +100,48 @@ export const findClosestDroppableTarget = (
   };
 };
 
+export const reparentInstanceMutable = (
+  instancesIndex: InstancesIndex,
+  instanceId: Instance["id"],
+  dropTarget: DroppableTarget
+) => {
+  const prevParent = instancesIndex.parentInstancesById.get(instanceId);
+  const nextParent = instancesIndex.instancesById.get(dropTarget.parentId);
+  const instance = instancesIndex.instancesById.get(instanceId);
+  if (
+    prevParent === undefined ||
+    nextParent === undefined ||
+    instance === undefined
+  ) {
+    return;
+  }
+
+  const prevPosition = prevParent.children.findIndex(
+    (child) => child.type === "instance" && child.id === instanceId
+  );
+  if (prevPosition === -1) {
+    return;
+  }
+
+  // if parent is the same, we need to adjust the position
+  // to account for the removal of the instance.
+  let nextPosition = dropTarget.position;
+  if (
+    nextPosition !== "end" &&
+    prevParent.id === nextParent.id &&
+    prevPosition < nextPosition
+  ) {
+    nextPosition -= 1;
+  }
+
+  prevParent.children.splice(prevPosition, 1);
+  if (nextPosition === "end") {
+    nextParent.children.push(instance);
+  } else {
+    nextParent.children.splice(nextPosition, 0, instance);
+  }
```

Comment:
It seems like this function makes passed `instancesIndex` invalid: modifies the tree, but doesn't update the Maps.

Currently this seem fine, because we create a fresh temporary index every time we apply the function. But I assume we'll have a re-usable index at some point.

Reply:
Yes, once we switch to normalized instances we gonna also modify instancesById map.

Reply:
Reusable index is recreated on every store update btw